sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala

```diff
@@ -3448,13 +3449,32 @@ object Sequence {
         || (estimatedStep == num.zero && start == stop),
       s"Illegal sequence boundaries: $start to $stop by $step")
 
-    val len = if (start == stop) 1L else 1L + (stop.toLong - start.toLong) / estimatedStep.toLong
-
-    require(
-      len <= MAX_ROUNDED_ARRAY_LENGTH,
-      s"Too long sequence: $len. Should be <= $MAX_ROUNDED_ARRAY_LENGTH")
-
-    len.toInt
+    try {
+      val delta = Math.subtractExact(stop.toLong, start.toLong)
+      if (delta == Long.MinValue && estimatedStep.toLong == -1L) {
+        // We must special-case division of Long.MinValue by -1 to catch potential unchecked
+        // overflow in next operation. Division does not have a builtin overflow check. We
+        // previously special-case div-by-zero.
+        throw new ArithmeticException("Long overflow (Long.MinValue / -1)")
+      }
+      val len = if (stop == start) 1L else Math.addExact(1L, (delta / estimatedStep.toLong))
+      if (len > MAX_ROUNDED_ARRAY_LENGTH) {
+        throw new IllegalArgumentException(s"Too long sequence: $len. Should be <= " +
```

Comment:
If the exception is an user-facing error, let's introduce an error class, and raise a Spark exception w/ it.

Reply:
Ah, I originally misunderstood your comment and with @ankurdave's help I learned about the errors defined in `QueryExecutionErrors.scala` and that they are actually used in this file. I thought `createArrayWithElementsExceedLimitError` might be the most fitting.